sympy/solvers/ode/riccati.py

```diff
@@ -0,0 +1,472 @@
+                r = (sgn*num1.subs(x, pole))/(den1.subs(x, pole))
+            else:
+                r = 0
+
+            # If multiplicity is 2, the coefficient to be added
+            # in the c-vector is c = (1 +- sqrt(1 + 4*r))/2
+            c[-1].extend([[(1 + sqrt(1 + 4*r))/2], [(1 - sqrt(1 + 4*r))/2]])
+
+        # Case 2
+        else:
+            # Generate the coefficients using the recurrence
+            # relation mentioned in (5.14) in the thesis (Pg 80)
+
+            # r_i = mul/2
+            ri = mul//2
+
+            # Find the Laurent series coefficients about the pole
+            ser = rational_laurent_series(num, den, x, pole, mul, 6)
+
+            # Start with an empty memo to store the coefficients
+            # This is for the plus case
+            temp = [0 for i in range(ri)]
+
+            # Base Case
+            temp[ri-1] = sqrt(ser[0])
+
+            # Iterate backwards to find all coefficients
+            for s in range(ri-1, 0, -1):
+                sm = 0
+                for j in range(s+1, ri):
+                    sm += temp[j-1]*temp[ri+s-j-1]
+                if s!= 1:
+                    temp[s-1] = (ser[mul-ri-s] - sm)/(2*temp[ri-1])
+
+            # Memo for the minus case
+            temp1 = list(map(lambda x: -x, temp))
+
+            # Find the 0th coefficient in the recurrence
+            temp[0] = (ser[mul-ri-s] - sm + ri*temp[ri-1])/(2*temp[ri-1])
+            temp1[0] = (ser[mul-ri-s] - sm + ri*temp1[ri-1])/(2*temp1[ri-1])
+
+            # Add both the plus and minus cases' coefficients
+            c[-1].extend([temp, temp1])
+    return c
+
+
+def construct_d(num, den, a, x, val_inf, mul):
+    N = -val_inf//2
+    ser = rational_laurent_series(num, den, x, oo, mul, 1)
+
+    # Case 4
+    if val_inf < 0:
+        temp = [0 for i in range(N+2)]
+        temp[N] = sqrt(ser[-mul - 1 + 2*N])
+        for s in range(N-1, -2, -1):
+            sm = 0
+            for j in range(s+1, N):
+                sm += temp[j]*temp[N+s-j]
+            if s != -1:
+                temp[s] = (ser[-mul - 1 + N+s] - sm)/(2*temp[N])
+        temp1 = list(map(lambda x: -x, temp))
+        temp[-1] = (ser[-mul - 1 + N+s] - temp[N] - sm)/(2*temp[N])
+        temp1[-1] = (ser[-mul - 1 + N+s] - temp1[N] - sm)/(2*temp1[N])
+        d = [temp, temp1]
+
+    # Case 5
+    elif val_inf == 0:
+        # List to store coefficients for plus case
+        temp = [0, 0]
+
+        # d_0  = sqrt(a_0)
+        temp[0] = sqrt(ser[-mul - 1])
+
+        # d_(-1) = a_(-1)/(2*d_0)
+        temp[-1] = ser[-mul - 2]/(2*temp[0])
+
+        # Coefficients for the minus case are just the negative
+        # of the coefficients for the positive case. Append both.
+        d = [temp, list(map(lambda x: -x, temp))]
+
+    # Case 6
+    else:
+        # s_oo = lim x->0 1/x**2 * a(1/x) which is equivalent to
+        # s_oo = lim x->oo x**2 * a(x)
+        s_inf = limit(x**2*a, x, oo)
```

Comment:
Is `x**2*a` a rational function here?

If so then we can compute the limit much more efficiently. If the numerator and the denominator have the same degree then it's just the ratio of their leading coefficients:
```python
In [6]: num = Poly(3*x**2 + x + 1)

In [7]: den = Poly(2*x**2 - x - 4)

In [8]: limit(num / den, x, oo)
Out[8]: 3/2

In [9]: num.LC() / den.LC()
Out[9]: 3/2
```
If the denominator has greater degree then it's zero. If the numerator has greater degree then it's just `oo*sign(num.LC() / den.LC())`.

Again a function could be made for this calculation.